saskatoon/unittests/test_permission.py

```diff
@@ -0,0 +1,44 @@
+import pytest
+from django.contrib.auth import get_user_model
+from pytest_django.asserts import assertRedirects
+from typing import Union
+
+STATIC_PAGES = ['calendar', 'property/create_public']
+STATIC_PAGES_PERMISSION_REQUIRED = ['harvest', 'property', 'organization', 'community', 'equipment', 'stats']
+
+User = get_user_model()
+
+
+def get_url(page_name: str) -> str:
+    return f"/{page_name}/"
+
+
+@pytest.mark.django_db
+def test_permission_static_pages_anonymous(client):
+    for page in STATIC_PAGES:
+        response = client.get(get_url(page))
+        assert response.status_code == 200, f"Page: {page}"
+
+    for page in STATIC_PAGES_PERMISSION_REQUIRED:
+        url = get_url(page)
+        response = client.get(url)
+        assertRedirects(response, f"/login/?next={url}", msg_prefix=f"Page: {page}")
+
+
+@pytest.mark.parametrize("auth_group, expected_status_codes",
+                         [('core', 200),
+                          ('pickleader', [200, 200, 200, 200, 200, 403]),
+                          ('volunteer', 403),
+                          ('owner', 403),
+                          ('contact', 403)])
```

Comment:
I understand that ideally a function should be written in such a way that it can be easily used without reading into its implementation. Meanwhile, as a function for testing, I also need to consider the maintenance cost (i.e. how easy to update it when the source codes change), for example, changing page name or auth group name or permission. That's why I want to centralize the page names into `STATIC_PAGES_PERMISSION_REQUIRED` (note that it is also used in `test_permission_static_pages_anonymous()`) and list permission by auth group. Since the implementation is only 7-line long, I think this trade-off is still acceptable.

I agree that it is confusing to allow `expected_status_codes` to be either a list or an int. I can make it always a list with 1 or n length as you suggested in the other comment.

What do you think?

Reply:
I understand your intentions but I would still prefer prioritizing clarity over cleverness.
If you want to maintain lists of pages to be reused across tests it totally makes sense but I would then suggest you create fixtures and pass them explicitly so there is no ambiguity.

Please also note that group names and `STAFF_GROUPS` are already defined here:
https://github.com/LesFruitsDefendus/saskatoon-ng/blob/4060b9e29d45bb4aeb7cc6e85e72ea9fee29aca5/saskatoon/member/models.py#L22
so if you're looking to be flexible to code changes then you should probably consider reusing these.

Reply:
>  I understand that ideally a function should be written in such a way that it can be easily used without reading into its implementation.

I just want to clarify that imo my initial comment was about more than readability, also maintainability. Suppose that for some reason someone inserts a new url into the `STATIC_PAGES_PERMISSION_REQUIRED` list but forgets to update the parametrization of this test. It is possible that this test will still pass even though we would no longer be testing the urls we thought we were and the parametrized status codes no longer are vis-à-vis the new list. What bothers me is not that we use a constant within a test. It is that the parametrized `expected_status_codes` are directly referring to that list and that is not immediately clear. I think that in general reading the function name, its parametrization (+ fixtures and docstring) should be enough to understand what the test is doing. 

Either way I understand an appreciate the concern for making code flexible and easy to maintain. However I believe that tests fall into a different category and should be straightforward. I personally don't see an issue with having to update a test in the event that we do a major change in the User model or add a new auth_group.
And as we add more tests further down the line I don't want to worry about breaking existing tests if I the new tests need slightly different inputs.

Again I am not against centralization but in this case I think a dictionary or some other data structure would be better fitted than a plain list.

Reply:
No worries, I get your points :)

I restructured the codes to make it both centralized and readable (hopefully). The page names and permission matrix now are in one dict constant, and the testing functions are free from these mess.

I didn't use the `AUTH_GROUPS` or `STAFF_GROUPS` from `models.py` so that the tests can be somewhat shielded from model implementation detail, and I think it is still okay here as they are defined as constants at the beginning.

Let me know what you think.